Trigger Saving Answer when Toggling to Other Question in Submission

bivanalhar opened the pull request:
## Issue
Continuation of https://github.com/Coursemology/coursemology2/pull/7278, this issue is happens when the autosave is not yet triggered while the user has already switched to another question in stepped-format assessment. As such, the autosave won't be triggered and the answer is lost when the page is refreshed or the user navigates to another page.

## Solution
Under normal circumstances, there is a prompt showing up to user when there's unsaved answers. However, it would be better to trigger the saving when the user navigates to another question. Here, we trigger the autosaving when the user clicks another bubble/step to navigate to another question

https://github.com/Coursemology/coursemology2/assets/16359075/b9f126f2-8d88-46de-9927-5a17a0503bc7

bivanalhar committed:
fix(autosave): autosave not triggered upon step click

- when the other question's step is clicked before autosave is triggered, it won't save
- we attempted to trigger save answer when clicking out to other question

cysjonathan commented on client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/SubmissionEditForm.jsx:
```diff
@@ -538,44 +545,51 @@ const SubmissionEditForm = (props) => {
+                  </SvgIcon>
+                }
+                onClick={() => {
+                  handleSubmit((data) =>
```
We shouldn't need to call onSaveDraft unless the form is dirty / not yet saved.
i.e. Don't call when answer is clean, or when answer is already saved.

bivanalhar replied:
understood, and done

cysjonathan commented on client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/SubmissionEditForm.jsx:
```diff
@@ -538,44 +545,51 @@ const SubmissionEditForm = (props) => {
-    </Stepper>
-  );
+  const renderSteppedQuestions = () => {
+    const answerId = getCurrentlySavingAnswerId();
```
This `answerId = getCurrentlySavingAnswerId()` shouldn't be outside the loop outside of Stepper.
This setup forces any change in answerId will trigger a re-render of ALL the stepper bubbles.

Inside the `questionIds.map((id, index) => ...` we already have the associated `questionId`, so inside the `onClick` event we can obtain `questions[id].answerId` directly.

bivanalhar replied:
understood. I moved out the saving instruction into the function instead, so that the fetching of the answerId will be done only when user click on the bubble (so it won't get re-rendered)

bivanalhar committed:
fix(autosave): add constraint on saving on click away

- if the currently displayed answer is both unsaved and dirty, proceed the autosave
- drag out the saving component into the function definition

bivanalhar committed:
fix(savingStatus): fix status on stale answer

- if failed due to stale answer, revert saving status to its previous state
- instead of keeping it on SAVING

ekowidianto commented on client/app/bundles/course/assessment/submission/reducers/answerFlags/index.ts:
```diff
@@ -58,11 +58,8 @@ export const answerFlagsSlice = createSlice({
-        isStaleAnswer?: boolean;
       }>,
     ) => {
-      if (action.payload.isStaleAnswer) return;
```
Why is this removed?

bivanalhar replied:
the previous notion of this stale answer is only when there's a race condition within the same action (for example, while you're typing and then the API for saving is sent, and then when it goes back the redux store for clientVersion is updated -> another API call will be received and this answerFlag would be updated eventually).

However, with now the additional action that might trigger API call (clicking other bubble), having this line will possibly make the state to be consistent in "Saving" mode, even though the answer has been saved properly.

So my approach here is basically if the staleAnswer is met, then we revert the savingStatus to what it was before the saving takes place.. it might still give the savingStatus the persistent "Saving" state, but the occurences will be much less likely

cysjonathan replied:
> it might still give the savingStatus the persistent "Saving" state, but the occurences will be much less likely
1. I managed to reproduce a locked "Saving" state when switching between bubbles. How can we prevent this?
2. When switching bubbles, we should immediately flush the outstanding by dispatching a save (done), as well as cancel the debounced saved (not yet done).
